ts/src/coinmetro.ts

```diff
@@ -0,0 +1,1922 @@
+        //             "margin": false
+        //         },
+        //         {
+        //             "pair": "PERPUSD",
+        //             "precision": 5,
+        //             "margin": false
+        //         },
+        //         {
+        //             "pair": "YFIEUR",
+        //             "precision": 5,
+        //             "margin": false
+        //         },
+        //         ...
+        //     ]
+        //
+        // todo: check what is precision
+        return this.parseMarkets (response);
+    }
+
+    parseMarket (market): Market {
+        const id = this.safeString (market, 'pair');
+        const parsedMarketId = this.parseMarketId (id);
+        const baseId = this.safeString (parsedMarketId, 'baseId');
+        const quoteId = this.safeString (parsedMarketId, 'quoteId');
+        const base = this.safeCurrencyCode (baseId);
+        const quote = this.safeCurrencyCode (quoteId);
+        // todo: check
+        const basePrecisionAndLimits = this.parseMarketPrecisionAndLimits (baseId);
+        const quotePrecisionAndLimits = this.parseMarketPrecisionAndLimits (quoteId);
+        const margin = this.safeValue (market, 'margin', false);
+        const tradingFees = this.safeValue (this.fees, 'trading', {});
+        return this.safeMarketStructure ({
+            'id': id,
+            'symbol': base + '/' + quote,
+            'base': base,
+            'quote': quote,
+            'settle': undefined,
+            'baseId': baseId,
+            'quoteId': quoteId,
+            'settleId': undefined,
+            'type': 'spot',
+            'spot': true,
+            'margin': margin,
+            'swap': false,
+            'future': false,
+            'option': false,
+            'active': true,
+            'contract': false,
+            'linear': undefined,
+            'inverse': undefined,
+            'taker': this.safeNumber (tradingFees, 'taker'),
+            'maker': this.safeNumber (tradingFees, 'maker'),
+            'contractSize': undefined,
+            'expiry': undefined,
+            'expiryDatetime': undefined,
+            'strike': undefined,
+            'optionType': undefined,
+            'precision': {
+                'amount': basePrecisionAndLimits['precision'],
+                'price': undefined,
+                'base': basePrecisionAndLimits['precision'],
+                'quote': quotePrecisionAndLimits['precision'],
+            },
+            'limits': {
+                'leverage': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+                'amount': {
+                    'min': basePrecisionAndLimits['minLimit'],
+                    'max': undefined,
+                },
+                'price': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+                'cost': {
+                    'min': quotePrecisionAndLimits['minLimit'],
```

Comment:
Can we avoid these direct accesses? The exchange might return an outlier and it will crash the code

Reply:
These object are returned by the `parseMarketPrecisionAndLimits` method, which uses `this.safeObject`, `this.safeInteger` and `this.safeNumber`. I think we can use the direct access here. If you insist, I can add these `safe...` methods here to.

Reply:
yeah in that case it should be ok